src/states/abstract/State.js

```diff
@@ -0,0 +1,126 @@
+import videojs from 'video.js';
+
+export default class State {
+
+  constructor(player) {
+    this.player = player;
+  }
+
+  /*
+   * This is the only allowed way to perform state transitions. State transitions usually
+   * happen in player event handlers. They can also happen recursively in `init`. They
+   * should _not_ happen in `cleanup`.
+   */
+  transitionTo(NewState, ...args) {
+    const player = this.player;
+    const previousState = this;
+
+    previousState.cleanup();
+    const newState = new NewState(player);
+
+    player.ads._state = newState;
+    player.ads.debug(previousState.constructor.name + ' -> ' + newState.constructor.name);
+    newState.init(player, ...args);
+  }
+
+  /*
+   * Implemented by subclasses to provide initialization logic when transitioning
+   * to a new state.
+   */
+  init() {}
+
+  /*
+   * Implemented by subclasses to provide cleanup logic when transitioning
+   * to a new state.
+   */
+  cleanup() {}
+
+  /*
+   * Default event handlers. Different states can override these to provide behaviors.
+   */
+  onPlay() {}
+  onPlaying() {}
+  onEnded() {}
+  onAdsReady() {
+    videojs.log.warn('Unexpected adsready event');
+  }
+  onAdsError() {}
+  onAdsCanceled() {}
+  onAdTimeout() {}
+  onAdStarted() {}
+  onContentUpdate() {}
+  onContentResumed() {}
+  onContentEnded() {
+    videojs.log.warn('Unexpected contentended event');
+  }
+  onNoPreroll() {}
+  onNoPostroll() {}
+
+  /*
+   * Method handlers. Different states can override these to provide behaviors.
+   */
+  startLinearAdMode() {
+    videojs.log.warn('Unexpected startLinearAdMode invocation ' +
+      '(State via ' + this.constructor.name + ')');
```

Comment:
I would expect a message like `Unexpected startLinearAdMode invocation (via ContentPlayback state)` personally

Reply:
I wanted to include the name of the file that the log message is in to make it easier to debug.

Reply:
Maybe I wasn't clear, I just meant having state be last in the message is clearer 😆 
`videojs.log.warn('Unexpected startLinearAdMode invocation (via', this.constructor.name, ' state)')`

Reply:
Oh, I wrote "State" in there as the name of the file, because that is what this.constructor.name would be.

Reply:
yeah the order of the words. What do you mean by "ClassName via ClassName"?

Reply:
In these log messages I've been putting in parens `ClassNameThatLoggedThisMessage` and for this exception the intended reading is `ClassNameThatLoggedThisMessage by way of ConcreteClassThatInheritedTheBehavior`

Reply:
💡  Ah I see. Makes sense